### PR TITLE
Add the new React rules config

### PR DESCRIPTION
**Pull request by matuzalemsteles**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,7 +34,9 @@
 		"eslint-plugin-liferayportal": "^1.0.0",
 		"eslint-plugin-no-for-of-loops": "^1.0.0",
 		"eslint-plugin-no-only-tests": "^2.1.0",
-		"eslint-plugin-notice": "0.7.8"
+		"eslint-plugin-notice": "0.7.8",
+		"eslint-plugin-react": "^7.12.4",
+		"eslint-plugin-react-hooks": "^1.5.1"
```
Comment (matuzalemsteles) [inline package.json:39]: hey @wincent I've been thinking about maybe adding these dependencies in `devDependencies` and `peerDependencies` what do you think?

File: scripts/lint.js [hunk 2]
```diff
@@ -8,7 +8,7 @@ const child_process = require('child_process');
 const fs = require('fs');
 const path = require('path');
 
-const LINT_GLOBS = ['**/*.js', '.*.js'];
+const LINT_GLOBS = ['**/!(copyright)*.js', '.*.js'];
```
Comment (wincent) [inline scripts/lint.js:11]: Good idea. 👍

File: package.json (hunk 1 above)
Comment (wincent) [reply package.json:39]: My (possibly inaccurate) understanding of the use, based on [the docs](https://docs.npmjs.com/files/package.json) is:

* `dependencies` are for things that your package needs in order to work.
* `devDependencies` are for things that you need in order to develop your package.
* `peerDependencies` are how you declare compatibility with things you expect to be present in the environment your package is installed into.

So, for a developer tool like "eslint-config-liferay", these things are all "dependencies" because without them the package doesn't actually work. But "prettier", for example, is in "devDependencies" because you don't need it to *use* "eslint-config-liferay", only to *develop* it.

My understanding is that if you add "eslint-config-liferay" in "devDependencies", that *doesn't* meant that the things it declares as "dependencies" suddenly become non-dev dependencies for you; they still remain dev-dependencies. It seems that what we have here is a declaration of what this package needs to work, what we need to work on it, and what we expect it to work with.

I looked at [eslint-config-airbnb to see if they follow this pattern or not](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/package.json), and the answer is "both" 😂; that is, they have some dependencies in "dependencies", others in "devDependencies" *and* "peerDependencies", and others in just "devDependencies".

"eslint-config-google", on the other hand [only has eslint as a dependency](https://github.com/google/eslint-config-google/blob/master/package.json), and it declares it in "devDependencies"/"peerDependencies".

TL;DR: I think the way we have it is right, but I could easily be mistaken!

File: copyright.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /**
- * © 2017 Liferay, Inc. <https://liferay.com>
+ * © <%= YEAR %> Liferay, Inc. <https://liferay.com>
```
Comment (wincent) [inline copyright.js:2]: (Just thinking out loud here...)

This will default to current YEAR, right? According to the various places where we have written about this:

* [This loop post](https://loop.liferay.com/es/home/-/loop/feed/12995353) says it should be "the year that was entered when the file was created".
* [This wiki page](https://grow.liferay.com/share/Liferay+Policy+for+Marking+Our+Own+Source+Code) says it should be "year when this file was first created/published".
* [This comment](https://github.com/liferay/clay/pull/1451#issuecomment-456384580) says "the year to whatever year that file was in (if that’s too much work, 2014 as the year of the first commit should do)"

It's that last rule that made me put 2017 here, because 2017 was the year of the first commit to this repo. If we change this to `YEAR`, then it will do the right thing for *new* files, because they'll get the whatever the current year is at the time the file is created. I think that's right for this project, but only because all the files currently have the headers they need.

But for projects that have pre-existing code, *if* they copy this as an example, then they'd get "2019" added to all their files, including things created in prior years, and that would be wrong (for the same reason that 2017 would be wrong). That's a big "if", but I wonder if there is a way to make this more "fool-proof" for such projects. Any ideas?

**Review by wincent — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (matuzalemsteles) [reply package.json:39]: Yeah, what I thought about it and adding in `peerDependencies` we can let the devs keep those dependencies easily updated without us having to update so often, but the problem of that we have to warn people that they will have to install these dependencies for the package to work as expected (this is a negative point...) another point to add in `peerDependencies` which helps in "deduplicating" the same packages in a project since it will use the package version from the project. I find it a little difficult to happen for this case. 

I think keeping some packages in `devDependencies` is more of the packages we can use locally in the project, as we are doing here. I think it is the use case the other repositories. Let people keep the configuration packages up to date without having to wait for us to update. But I think we can go like this now and see if that's something we'll need in the future...

File: copyright.js (hunk 3 above)
Comment (matuzalemsteles) [reply copyright.js:2]: Good point about this, in fact I would expect the plugin would do this for us, according to the date of creation of the file, check the valid year for such. But it seems he does not do that 😞. That would be the only idea I would have... I do not know if we can do something like that on our side.

Comment (wincent) [reply copyright.js:2]: Yeah I think the problem is that Git doesn't manage creation dates, so if I clone a repo and then run the lint, "2019" will be used as creation date. Given that Git is used so pervasively, it's a pretty good argument against implementing the functionality in the plug-in.

File: package.json (hunk 1 above)
Comment (wincent) [reply package.json:39]: It's definitely a trade-off. Using `peerDependencies` more heavily means:

- Good: Greater control for the user, and the ability to update dependencies pre-emptively without waiting for "eslint-config-liferay" to update them.
- Bad: Less of a batteries-included approach to linting.

I think we should bias towards being a "batteries-included" config, because:

- Ultimately the "people" using this config are actually just us, which means that if we ever need to update the config, nothing will block us.
- We can provide a single, standard linting configuration that we use across all of our projects that is not dependent on whether a particular project has choosen to update its `peerDependencies` or not; you'll just know that if you update to the latest "eslint-config-liferay", you'll always get the current "best practice".
- The ability to update to new versions of dependencies via `peerDependencies` is likely overrated, especially if we use looser ranges (like: `^0.7.8` instead of `0.7.8`).
- Package deduplication is likely a non-issue not just because we have large disks and fast computers, but also partly because we don't expect people to be installing a bunch of competing lint configurations that might specify other versions of our dependencies (this due the package being a "batteries-included" thing), and partly because these dev-time dependencies can't have any negative impact on production-facing things like bundle sizes and so forth.

Comment (matuzalemsteles) [reply package.json:39]: Good points I think it is really better to go without `peerDependencies` for our case. 🙂

File: copyright.js (hunk 3 above)
Comment (matuzalemsteles) [reply copyright.js:2]: Yeah, I do not know if it would help a lot, but maybe we can add warnings about using copyright properly by referring to posts. In the documentation here or README.md, where to become more visible.

**Comment by wincent** [comment]
Cherry-picked some bits of this over into #42 as I explained over there. The only bit I didn't cherry-pick was "Change the copyright template to accept param <%= YEAR %>" because it seemed unclear whether it was correct based on our conversation in [this thread](https://github.com/liferay/eslint-config-liferay/pull/37#discussion_r266415163).

I chose the rules based on running my PR against a React project in liferay-portal (segments-web); basically shooting for a conservative subset of the "react/recommended" config that wouldn't be too noisy for the initial roll out. Over time we can try to make the configs converge between here, Clay, and liferay-portal, so that we can just define it here in one place and the same rules can be used everywhere.
